enhancements/machine-api/aws-placement-groups.md

```diff
@@ -0,0 +1,543 @@
+  // AWSPartitionPlacementGroupType is the "Partition" placement group type.
+  // Partition placement groups reduce the likelihood of hardware failures
+  // disrupting your application's availability.
+  // Partition placement groups are recommended for use with large scale
+  // distributed and replicated workloads.
+  AWSPartitionPlacementGroupType AWSPlacementGroupType = "Partition"
+  // AWSSpreadPlacementGroupType is the "Spread" placement group type.
+  // Spread placement groups place instances on distinct racks within the availability
+  // zone. This ensures instances each have their own networking and power source
+  // for maximum hardware fault tolerance.
+  // Spread placement groups are recommended for a small number of critical instances
+  // which must be kept separate from one another.
+  // Using a Spread placement group imposes a limit of seven instances within
+  // the placement group within a single availability zone.
+  AWSSpreadPlacementGroupType    AWSPlacementGroupType = "Spread"
+)
+
+// AWSPlacementGroupUnion is a discriminated union of placement group configuration.
+// +union
+type AWSPlacementGroupUnion struct {
+  // GroupType specifies the type of AWS placement group to use for this Machine.
+  // This parameter is only used when a Machine is being created and the named
+  // placement group does not exist.
+  // Valid values are "Cluster", "Partition", "Spread".
+  // If no value is given and a placement group is required to be created, this is
+  // considered an error.
+  // Note: If the value of this field is "Spread", instanceTenancy may not be set
+  // to "dedicated".
+  // +kubebuilder:validation:Enum:="Cluster";"Partition";"Spread"
+  // +unionDiscriminator
+  // +kubebuilder:validation:Required
+  // +required
+  GroupType AWSPlacementGroupType `json:"groupType,omitempty"`
+
+  // +optional
+  Cluster *AWSClusterPlacement `json:"cluster,omitempty"`
+  // +optional
+  Partition *AWSPartitionPlacement `json:"partition,omitempty"`
+  // +optional
+  Spread *AWSSpreadPlacement `json:"spread,omitempty"`
+}
+
+// AWSClusterPlacement defines the configuration for cluster placement groups.
+type AWSClusterPlacement struct {}
+
+// AWSPartitionPlacement defines the configuration for partition placement groups.
+type AWSPartitionPlacement struct {
+  // Count specifies the number of partitions for a Partition placement
+  // group. This value is only observed when creating a placement group and
+  // only when the `groupType` is set to `Partition`.
+  // Note the partition count of a placement group cannot be changed after creation.
+  // If unset, AWS will provide a default partition count.
+  // This default is currently 2.
+  // Note: If instanceTenancy is set to "dedicated", the maximum number of partitions
+  // is limited to 2.
+  // +kubebuilder:validation:Minimum:=1
+  // +kubebuilder:validation:Maximum:=7
+  // +optional
+  Count int32 `json:"count,omitempty"`
```

Comment:
What happens if different machinesets specify this number differently for the same GroupName?

Reply:
> The more I read and come to understand how placement groups work, the more it seems to me that placement groups really should be configured as separate config objects from machines/machinesets. Defining placement groups within machinesets makes the semantics extremely confusing. As it is, both the basic usage and behaviors of placement groups as well as edge cases (such as the ones David brings up) are difficult to understand.

I could envisage a separate CRD for this that then gets created, thought that's a lot of work and means more controllers and more complication from a user perspective, so I think there's a balance to make here.

The upstream community in Azure already have similar concepts being created by the Machine controller so we have followed that pattern for now.

> What happens if different machinesets specify this number differently for the same GroupName?

We can validate that this configuration is matching the existing placement group and error out if it isn't, this would enforce the count and type to be consistent across multiple MachineSets